README
NAME
    Pod::Usage::Return - pod2usage that returns instead of exits

VERSION
    version 0.004

SYNOPSIS
        use Pod::Usage::Return qw( pod2usage );

        exit pod2usage(0);

        sub main {
            return pod2usage("ERROR: An error occurred!") if $ERROR;
        }

        exit pod2usage( -exitval => 1, -message => 'ERROR: An error occurred' );

DESCRIPTION
    This is a drop-in replacement for Pod::Usage "pod2usage" that returns
    the exit value instead of calling exit.

RATIONALE
    Testing that your command-line script works is a good thing. It's a lot
    easier to test a module, so writing your command-line script as a module
    ("modulino") makes it easier to test.

    Unfortunately, Pod::Usage automatically calls "exit", which again makes
    it harder to test your script. There is a way to prevent Pod::Usage from
    exiting, but it makes using Pod::Usage a lot less convenient.

    This module provides a drop-in "pod2usage" replacement that returns the
    exit code instead of exiting, so that you can easily test your script
    while using Pod::Usage.

FUNCTIONS
  pod2usage
    See Pod::Usage for documentation. Returns the exit code instead of
    calling exit().

AUTHOR
    Doug Bell <preaction@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Doug Bell.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
Changelog for Pod-Usage-Return

0.004 2014-11-03T00:39:39
 - fix prereqs
 - add travis-ci
 - SurgicalPodWeaver -> PodWeaver

0.003 2014-06-28T06:21:53
 - release v0.003
 - add rationale
 - fix misspelling in POD

0.002 2014-05-10T02:39:43
 - update build artifacts
 - tests require Capture::Tiny
 - add a minimum version for Pod::Usage

0.001 2014-05-09T04:05:38
 - initial build artifacts
 - initial pod2usage use-cases



Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.023.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.28"
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build" => "0.28"
  },
  "dist_abstract" => "pod2usage that returns instead of exits",
  "dist_author" => [
    "Doug Bell <preaction\@cpan.org>"
  ],
  "dist_name" => "Pod-Usage-Return",
  "dist_version" => "0.004",
  "license" => "perl",
  "module_name" => "Pod::Usage::Return",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Exporter" => 0,
    "Pod::Usage" => "1.36",
    "perl" => "5.008",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "Capture::Tiny" => 0,
    "Test::Compile" => 0,
    "Test::More" => 0
  }
);


my %fallback_build_requires = (
  "Capture::Tiny" => 0,
  "Module::Build" => "0.28",
  "Test::Compile" => 0,
  "Test::More" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Pod/Usage/Return.pm
package Pod::Usage::Return;
# ABSTRACT: pod2usage that returns instead of exits

use strict;
use warnings;
use Pod::Usage ();
use base 'Exporter';
our @EXPORT = qw( pod2usage );

sub pod2usage {
    my $exitval = 2;
    my %args;
    if ( @_ == 1 ) {
        if ( ref $_[0] eq 'HASH' ) {
            %args = %{$_[0]};
        }
        elsif ( $_[0] =~ /^\d+$/ ) {
            $exitval = $_[0];
        }
        else {
            $args{-message} = $_[0];
            $args{-verbose} = 0;
        }
    }
    else {
        %args = @_;
    }
    $args{-exitval} = 'NOEXIT';
    if ( $exitval >= 2 ) {
        $args{-output} = *STDERR;
    }
    Pod::Usage::pod2usage( \%args );
    return $exitval;
}

1;
__END__

=head1 SYNOPSIS

    use Pod::Usage::Return qw( pod2usage );

    exit pod2usage(0);

    sub main {
        return pod2usage("ERROR: An error occurred!") if $ERROR;
    }

    exit pod2usage( -exitval => 1, -message => 'ERROR: An error occurred' );

=head1 DESCRIPTION

This is a drop-in replacement for L<Pod::Usage> C<pod2usage> that returns the
exit value instead of calling exit.

=head1 RATIONALE

Testing that your command-line script works is a good thing. It's a lot easier
to test a module, so writing your command-line script as a module ("modulino")
makes it easier to test.

Unfortunately, L<Pod::Usage> automatically calls C<exit>, which again makes it
harder to test your script. There is a way to prevent Pod::Usage from exiting,
but it makes using Pod::Usage a lot less convenient.

This module provides a drop-in C<pod2usage> replacement that returns the exit
code instead of exiting, so that you can easily test your script while using
Pod::Usage.

=head1 FUNCTIONS

=head2 pod2usage

See L<Pod::Usage> for documentation. Returns the exit code instead of calling
exit().


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.023.
use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "pod2usage that returns instead of exits",
  "AUTHOR" => "Doug Bell <preaction\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.28"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build" => "0.28"
  },
  "DISTNAME" => "Pod-Usage-Return",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008",
  "NAME" => "Pod::Usage::Return",
  "PREREQ_PM" => {
    "Exporter" => 0,
    "Pod::Usage" => "1.36",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Capture::Tiny" => 0,
    "Test::Compile" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.004",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Capture::Tiny" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "Module::Build" => "0.28",
  "Pod::Usage" => "1.36",
  "Test::Compile" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



